subprojects/jacoco/src/main/java/org/gradle/internal/jacoco/transform/AgentJarExtraction.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.jacoco.transform;
+
+import org.gradle.api.GradleException;
+import org.gradle.api.UncheckedIOException;
+import org.gradle.api.artifacts.transform.InputArtifact;
+import org.gradle.api.artifacts.transform.TransformAction;
+import org.gradle.api.artifacts.transform.TransformOutputs;
+import org.gradle.api.artifacts.transform.TransformParameters;
+import org.gradle.api.file.FileSystemLocation;
+import org.gradle.api.provider.Provider;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.Enumeration;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+public abstract class AgentJarExtraction implements TransformAction<TransformParameters.None> {
+
+    @InputArtifact
+    public abstract Provider<FileSystemLocation> getInputArtifact();
+
+    @Override
+    public void transform(TransformOutputs outputs) {
+        File input = getInputArtifact().get().getAsFile();
+        File output = outputs.file(input.getName());
+
+        extractAgentJar(input, output);
+    }
+
+    private void extractAgentJar(File input, File output) {
+        try (ZipFile zipFile = new ZipFile(input)) {
+            Enumeration<? extends ZipEntry> zipEntries = zipFile.entries();
+            while (zipEntries.hasMoreElements()) {
+                ZipEntry zipEntry = zipEntries.nextElement();
+                if (zipEntry.getName().endsWith("jacocoagent.jar")) {
+                    try (InputStream in = zipFile.getInputStream(zipEntry)) {
+                        Files.copy(in, output.toPath(), StandardCopyOption.REPLACE_EXISTING);
+                    }
+                    return;
+                }
+            }
```

Comment:
Should this transform fail if it can't find the agent jar inside the library jar?

Reply:
There will be a failure later which is probably fine. Maybe you can check if that failure is informative enough?
Ideally we would also have a test case for the failure case. We have test fixtures to create repositories with modules. You could use that to have a repository with an empty `jacoco` jar.

Reply:
We added a failure message and a test for this.